### DISCUSSION TITLE
Determine if an object is "new" in the current session or not

### DISCUSSION THREAD

**Discussion by dkbarn** [sqlalchemy/sqlalchemy#10475] [Usage Questions]
Given an ORM object, I'm looking for a way of being able to determine whether that object originated from a database query or was newly created in the active session.

The inspection system offers a way to inspect the state of an ORM object, using ```inspect(obj).persistent``` for example, but this is not sufficient. If a flush happened to occur on the current session, then an ORM object will be marked as persistent even though it was newly created in this session, therefore it is not distinguishable from objects that were fetched from a query and existed in the database prior to the current session.

Is there a way to make this distinction?

**Comment by zzzeek** [comment]
hi -

to tag persistent to the lifespan of a session you would want to use [lifecycle events](https://docs.sqlalchemy.org/en/20/orm/session_events.html#object-lifecycle-events) to track the information you'd need.

below is a simplistic hook that adds objects to a set that's stored in [session.info](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.info) which records all the "pending to persistent" objects:


```py

from sqlalchemy.orm import sessionmaker

@event.listens_for(sessionmaker, "pending_to_persistent")
def intercept_pending_to_persistent(session, object_):
     session.info.setdefault("became_persistent", set()).add(object_)
```
